src/libraries/System.Linq/src/System/Linq/OrderBy.cs

```diff
@@ -132,6 +137,17 @@ public static partial class Enumerable
 
             return source.CreateOrderedEnumerable(keySelector, comparer, true);
         }
+
+        /// <summary>Gets whether the results of an unstable sort will be observably the same as a stable sort.</summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static bool TypeIsImplicitlyStable<T>() =>
```

Comment:
For my education, how does stability become a concern in floating point numbers?

Reply:
In particular NaNs can compare as equal even if their bit representations differ, in which case resulting order could be observable.  I don't know that it would matter here, but I didn't want to risk it.  @tannergooding?

Reply:
My goal around this in this PR was to not make observable changes to the resulting ordering.

Reply:
> Could we possibly extend this optimization to other structs?

We could, but there's no way to do that as a JIT intrinsic to my knowledge, so we'd need to do something like create a static readonly bool to cache the answer for the TKey, and then use reflection to evaluate the tuple to see whether all of the elements are supported.